File: java/src/irrgarten/Shield.java

package irrgarten;

/**
 * @brief Representa un escudo en el laberinto a partir de la clase `CombatElement`.
 */
public class Shield extends CombatElement {
   
    /**
     * @brief Constructor de la clase `Shield`.
     * @param protection Nivel de protección que ofrece el escudo.
     * @param uses Número inicial de usos del escudo.
     */
    public Shield(float protection, int uses) {
        super(protection, uses); // Llama al constructor de la clase CombatElement.
    }

    /**
     * @brief Proporciona protección al jugador y reduce los usos del escudo.
     * @return El nivel de protección del escudo si tiene usos restantes, o 0 si no tiene usos.
     */
    public float protect() {
        return super.produceEffect();
    }

    /**
     * @brief Representa el escudo como una cadena de texto.
     * @return Una cadena que contiene el nivel de protección y los usos restantes del escudo.
     */
    @Override
    public String toString() {
        return ("S"+super.toString()); 
    }

    /**
     * @brief Determina si el escudo debe ser descartado.
     * @return `true` si el escudo debe ser descartado, `false` en caso contrario.
     */
    public boolean discard() {
        return super.discard();
    }
}

File: ruby/Irrgarten/Shield.rb
# encoding : UTF−8

class Shield

  def initialize(protection, uses)
    @protection = protection.tp_f
    @uses = uses.to_i
  end

  def attack
    if @uses > 0
      @uses -= 1
      return @@protection
    else return 0
    end
  end

  def to_s
    "W[#{@protection}, #{@uses}]"
  end
  
  def discard
    Dice.discard_element(@uses)
  end
  
end

